### ISSUE TITLE
Undefined defaultValue

### ISSUE THREAD

**Issue by glutentag** [sequelize/sequelize#5467]
Hello,

I have found an issue while using default values. If we create a model with an undefined column with a default value, like
```javascript
Model.build({
    number1: undefined
});
```
the instance still has an undefined value.

This is due to instance.initValues logic, which set the default values when undefined, then set all the values. A defined undefined override the default value.
```javascript
  if (Object.keys(defaults).length) {
      for (key in defaults) {
        if (values[key] === undefined) {
          this.set(key, Utils.toDefaultValue(defaults[key]), defaultsOptions);
        }
      }
    }
  }
  this.set(values, options);
```
It can be fixed with a delete values[key] after setting the default value.

I have 2 questions regarding default values though:

- Shouldn't the default value be also applied when having a null value?
- The default values are applied only with a new record. Shouldn't it be applied also when updating a record?

**Comment by mickhansen** [comment]
What exactly are you expecting to happen?
undefined is undefined - It's not the same as null.
I would personally expect a default value to be applied if i passed undefined.

**Comment by mickhansen** [comment]
Ignore my previous (now deleted comment), it's early here and i can't read apparently.
Passing `undefined`  should absolutely still apply the default value.
The default value should not be provided when passing `null`, `null` is different from undefined in javascript, and passing null for the DB is a clear intent.

**Comment by mickhansen** [comment]
Merged #5468

**Comment by glutentag** [comment]
No problem.
After exchanging with my team, we agreed sequelize's current behaviour is the same than for any DB, the special need we have should be defined in the column's setter.
